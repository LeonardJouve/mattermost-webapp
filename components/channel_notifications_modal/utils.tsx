// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {FormattedMessage} from 'react-intl';

import React from 'react';

import {t} from 'utils/i18n';
import {FieldsetCheckbox} from 'components/widgets/modals/generic/checkbox-item-creator';

import {FieldsetRadio} from '../widgets/modals/generic/radio-item-creator';
import {NotificationLevels} from '../../utils/constants';
import {ChannelNotifyProps} from '@mattermost/types/channels';
import {UserNotifyProps} from '@mattermost/types/users';

export type ChannelMemberNotifyProps = Partial<ChannelNotifyProps> & Pick<UserNotifyProps, 'desktop_threads' | 'push_threads'>

export const MuteAndIgnoreSectionTitle = {
    id: t('channel_notifications.muteAndIgnore'),
    defaultMessage: 'Mute or ignore',
};

export const NotifyMeTitle = {
    id: t('channel_notifications.NotifyMeTitle'),
    defaultMessage: 'Notify me about…',
};

export const ThreadsReplyTitle = {
    id: t('channel_notifications.ThreadsReplyTitle'),
    defaultMessage: 'Thread reply notifications',
};

export const DesktopNotificationsSectionTitle = {
    id: t('channel_notifications.desktopNotificationsTitle'),
    defaultMessage: 'Desktop Notifications',
};

export const DesktopNotificationsSectionDesc = {
    id: t('channel_notifications.desktopNotificationsDesc'),
    defaultMessage: 'Available on Chrome, Edge, Firefox, and the Mattermost Desktop App.',
};

export const MobileNotificationsSectionTitle = {
    id: t('channel_notifications.mobileNotificationsTitle'),
    defaultMessage: 'Mobile Notifications',
};

export const MobileNotificationsSectionDesc = {
    id: t('channel_notifications.mobileNotificationsDesc'),
    defaultMessage: 'Notification alerts are pushed to your mobile device when there is activity in Mattermost.',
};

export const MuteChannelDesc = {
    id: t('channel_notifications.muteChannelDesc'),
    defaultMessage: 'Turns off notifications for this channel. You’ll still see badges if you’re mentioned.',
};
export const IgnoreMentionsDesc = {
    id: t('channel_notifications.ignoreMentionsDesc'),
    defaultMessage: 'When enabled, @channel, @here and @all will not trigger mentions or mention notifications in this channel',
};

export const MuteChannelInputFieldData: FieldsetCheckbox = {
    title: {
        id: t('channel_notifications.muteChannelTitle'),
        defaultMessage: 'Mute channel',
    },
    name: 'mute channel',
    dataTestId: 'muteChannel',
};

export const DesktopReplyThreadsInputFieldData: FieldsetCheckbox = {
    title: {
        id: t('channel_notifications.checkbox.threadsReplyTitle'),
        defaultMessage: 'Notify me about replies to threads I’m following',
    },
    name: 'desktop reply threads',
    dataTestId: 'desktopReplyThreads',
};

export const MobileReplyThreadsInputFieldData: FieldsetCheckbox = {
    title: {
        id: t('channel_notifications.checkbox.threadsReplyTitle'),
        defaultMessage: 'Notify me about replies to threads I’m following',
    },
    name: 'mobile reply threads',
    dataTestId: 'mobileReplyThreads',
};

export const sameMobileSettingsDesktopInputFieldData: FieldsetCheckbox = {
    title: {
        id: t('channel_notifications.checkbox.sameMobileSettingsDesktop'),
        defaultMessage: 'Use the same notification settings as desktop',
    },
    name: 'same mobile settings as Desktop',
    dataTestId: 'sameMobileSettingsDesktop',
};

export const IgnoreMentionsInputFieldData: FieldsetCheckbox = {
    title: {
        id: t('channel_notifications.ignoreMentionsTitle'),
        defaultMessage: 'Ignore mentions for @channel, @here and @all',
    },
    name: 'ignore mentions',
    dataTestId: 'ignoreMentions',
};

export const AutoFollowThreadsTitle = {
    id: t('channel_notifications.autoFollowThreadsTitle'),
    defaultMessage: 'Follow all threads in this channel',
};

export const AutoFollowThreadsDesc = {
    id: t('channel_notifications.autoFollowThreadsDesc'),
    defaultMessage: 'When enabled, all new replies in this channel will be automatically followed and will appear in your Threads view.',
};

export const AutoFollowThreadsInputFieldData: FieldsetCheckbox = {
    title: {
        id: t('channel_notifications.checkbox.autoFollowThreadsTitle'),
        defaultMessage: 'Automatically follow threads in this channel',
    },
    name: 'auto follow threads',
    dataTestId: 'autoFollowThreads',
};

export const desktopNotificationInputFieldData = (defaultOption: string): FieldsetRadio => {
    return {
        options: [
            {
                dataTestId: `desktopNotification-${NotificationLevels.ALL}`,
                title: {
                    id: 'channel_notifications.desktopNotificationAllLabel',
                    defaultMessage: 'All new messages',
                },
                name: `desktopNotification-${NotificationLevels.ALL}`,
                key: `desktopNotification-${NotificationLevels.ALL}`,
                value: NotificationLevels.ALL,
                suffix: defaultOption === NotificationLevels.ALL ? (
                    <FormattedMessage
                        id='channel_notifications.default'
                        defaultMessage='(default)'
                    />) : undefined,
            },
            {
                dataTestId: `desktopNotification-${NotificationLevels.MENTION}`,
                title: {
                    id: 'channel_notifications.desktopNotificationMentionLabel',
                    defaultMessage: 'Mentions, direct messages, and keywords only',
                },
                name: `desktopNotification-${NotificationLevels.MENTION}`,
                key: `desktopNotification-${NotificationLevels.MENTION}`,
                value: NotificationLevels.MENTION,
                suffix: defaultOption === NotificationLevels.MENTION ? (
                    <FormattedMessage
                        id='channel_notifications.default'
                        defaultMessage='(default)'
                    />) : undefined,
            },
            {
                dataTestId: `desktopNotification-${NotificationLevels.NONE}`,
                title: {
                    id: 'channel_notifications.desktopNotificationNothingLabel',
                    defaultMessage: 'Nothing',
                },
                name: `desktopNotification-${NotificationLevels.NONE}`,
                key: `desktopNotification-${NotificationLevels.NONE}`,
                value: NotificationLevels.NONE,
                suffix: defaultOption === NotificationLevels.NONE ? (
                    <FormattedMessage
                        id='channel_notifications.default'
                        defaultMessage='(default)'
                    />) : undefined,
            },
        ],
    };
};

export const mobileNotificationInputFieldData = (defaultOption: string): FieldsetRadio => {
    return {
        options: [
            {
                dataTestId: `MobileNotification-${NotificationLevels.ALL}`,
                title: {
                    id: 'channel_notifications.MobileNotificationAllLabel',
                    defaultMessage: 'All new messages',
                },
                name: `MobileNotification-${NotificationLevels.ALL}`,
                key: `MobileNotification-${NotificationLevels.ALL}`,
                value: NotificationLevels.ALL,
                suffix: defaultOption === NotificationLevels.ALL ? (
                    <FormattedMessage
                        id='channel_notifications.default'
                        defaultMessage='(default)'
                    />) : undefined,
            },
            {
                dataTestId: `MobileNotification-${NotificationLevels.MENTION}`,
                title: {
                    id: 'channel_notifications.MobileNotificationMentionLabel',
                    defaultMessage: 'Mentions, direct messages, and keywords only',
                },
                name: `MobileNotification-${NotificationLevels.MENTION}`,
                key: `MobileNotification-${NotificationLevels.MENTION}`,
                value: NotificationLevels.MENTION,
                suffix: defaultOption === NotificationLevels.MENTION ? (
                    <FormattedMessage
                        id='channel_notifications.default'
                        defaultMessage='(default)'
                    />) : undefined,
            },
            {
                dataTestId: `MobileNotification-${NotificationLevels.NONE}`,
                title: {
                    id: 'channel_notifications.MobileNotificationNothingLabel',
                    defaultMessage: 'Nothing',
                },
                name: `MobileNotification-${NotificationLevels.NONE}`,
                key: `MobileNotification-${NotificationLevels.NONE}`,
                value: NotificationLevels.NONE,
                suffix: defaultOption === NotificationLevels.NONE ? (
                    <FormattedMessage
                        id='channel_notifications.default'
                        defaultMessage='(default)'
                    />) : undefined,
            },
        ],
    };
};

