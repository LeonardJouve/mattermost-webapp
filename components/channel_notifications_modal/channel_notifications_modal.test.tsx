// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {ComponentProps} from 'react';
import {shallow} from 'enzyme';

import {IgnoreChannelMentions, NotificationLevels} from 'utils/constants';
import {TestHelper} from 'utils/test_helper';

import ChannelNotificationsModal from 'components/channel_notifications_modal/channel_notifications_modal';

import {UserNotifyProps} from '@mattermost/types/users';
import {ChannelMembership} from '@mattermost/types/channels';

describe('components/channel_notifications_modal/ChannelNotificationsModal', () => {
    const baseProps: ComponentProps<typeof ChannelNotificationsModal> = {
        onExited: jest.fn(),
        channel: TestHelper.getChannelMock({
            id: 'channel_id',
            display_name: 'channel_display_name',
        }),
        channelMember: {
            notify_props: {
                desktop: NotificationLevels.ALL,
                mark_unread: NotificationLevels.ALL,
                push: NotificationLevels.DEFAULT,
                ignore_channel_mentions: IgnoreChannelMentions.DEFAULT,
                desktop_threads: NotificationLevels.ALL,
                push_threads: NotificationLevels.DEFAULT,
            },
        } as unknown as ChannelMembership,
        currentUser: TestHelper.getUserMock({
            id: 'current_user_id',
            notify_props: {
                desktop: NotificationLevels.ALL,
                desktop_threads: NotificationLevels.ALL,
            } as UserNotifyProps,
        }),
        sendPushNotifications: true,
        actions: {
            updateChannelNotifyProps: jest.fn(),
        },
    };

    test('should match snapshot', () => {
        const wrapper = shallow(
            <ChannelNotificationsModal {...baseProps}/>,
        );

        expect(wrapper).toMatchSnapshot();
    });
});
