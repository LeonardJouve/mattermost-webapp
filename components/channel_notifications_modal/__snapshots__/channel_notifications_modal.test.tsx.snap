// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/channel_notifications_modal/ChannelNotificationsModal should match snapshot 1`] = `
<Modal
  animation={true}
  aria-labelledby="channelNotificationModalLabel"
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogClassName="a11y__modal channel-notifications-settings-modal"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[Function]}
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  role="dialog"
  show={true}
>
  <ModalHeader
    handleClose={[Function]}
    id="channelNotificationModalLabel"
    subtitle="channel_display_name"
    title="Notification Preferences"
  />
  <main
    className="channel-notifications-settings-modal__body"
  >
    <div
      className="channel-notifications-settings-modal__content"
    >
      <SectionCreator
        content={
          <React.Fragment>
            <CheckboxItemCreator
              description={
                Object {
                  "defaultMessage": "Turns off notifications for this channel. You’ll still see badges if you’re mentioned.",
                  "id": "channel_notifications.muteChannelDesc",
                }
              }
              handleChange={[Function]}
              inputFieldData={
                Object {
                  "dataTestId": "muteChannel",
                  "name": "mute channel",
                  "title": Object {
                    "defaultMessage": "Mute channel",
                    "id": "channel_notifications.muteChannelTitle",
                  },
                }
              }
              inputFieldValue={false}
            />
            <CheckboxItemCreator
              description={
                Object {
                  "defaultMessage": "When enabled, @channel, @here and @all will not trigger mentions or mention notifications in this channel",
                  "id": "channel_notifications.ignoreMentionsDesc",
                }
              }
              handleChange={[Function]}
              inputFieldData={
                Object {
                  "dataTestId": "ignoreMentions",
                  "name": "ignore mentions",
                  "title": Object {
                    "defaultMessage": "Ignore mentions for @channel, @here and @all",
                    "id": "channel_notifications.ignoreMentionsTitle",
                  },
                }
              }
              inputFieldValue={false}
            />
          </React.Fragment>
        }
        title={
          Object {
            "defaultMessage": "Mute or ignore",
            "id": "channel_notifications.muteAndIgnore",
          }
        }
      />
      <div
        className="channel-notifications-settings-modal__divider"
      />
      <SectionCreator
        content={
          <React.Fragment>
            <RadioItemCreator
              handleChange={[Function]}
              inputFieldData={
                Object {
                  "options": Array [
                    Object {
                      "dataTestId": "desktopNotification-all",
                      "key": "desktopNotification-all",
                      "name": "desktopNotification-all",
                      "suffix": <Memo(MemoizedFormattedMessage)
                        defaultMessage="(default)"
                        id="channel_notifications.default"
                      />,
                      "title": Object {
                        "defaultMessage": "All new messages",
                        "id": "channel_notifications.desktopNotificationAllLabel",
                      },
                      "value": "all",
                    },
                    Object {
                      "dataTestId": "desktopNotification-mention",
                      "key": "desktopNotification-mention",
                      "name": "desktopNotification-mention",
                      "suffix": undefined,
                      "title": Object {
                        "defaultMessage": "Mentions, direct messages, and keywords only",
                        "id": "channel_notifications.desktopNotificationMentionLabel",
                      },
                      "value": "mention",
                    },
                    Object {
                      "dataTestId": "desktopNotification-none",
                      "key": "desktopNotification-none",
                      "name": "desktopNotification-none",
                      "suffix": undefined,
                      "title": Object {
                        "defaultMessage": "Nothing",
                        "id": "channel_notifications.desktopNotificationNothingLabel",
                      },
                      "value": "none",
                    },
                  ],
                }
              }
              inputFieldValue="all"
              title={
                Object {
                  "defaultMessage": "Notify me about…",
                  "id": "channel_notifications.NotifyMeTitle",
                }
              }
            />
          </React.Fragment>
        }
        description={
          Object {
            "defaultMessage": "Available on Chrome, Edge, Firefox, and the Mattermost Desktop App.",
            "id": "channel_notifications.desktopNotificationsDesc",
          }
        }
        title={
          Object {
            "defaultMessage": "Desktop Notifications",
            "id": "channel_notifications.desktopNotificationsTitle",
          }
        }
        titleSuffix={<React.Fragment />}
      />
      <div
        className="channel-notifications-settings-modal__divider"
      />
      <SectionCreator
        content={
          <React.Fragment>
            <CheckboxItemCreator
              handleChange={[Function]}
              inputFieldData={
                Object {
                  "dataTestId": "sameMobileSettingsDesktop",
                  "name": "same mobile settings as Desktop",
                  "title": Object {
                    "defaultMessage": "Use the same notification settings as desktop",
                    "id": "channel_notifications.checkbox.sameMobileSettingsDesktop",
                  },
                }
              }
              inputFieldValue={false}
            />
            <React.Fragment>
              <RadioItemCreator
                handleChange={[Function]}
                inputFieldData={
                  Object {
                    "options": Array [
                      Object {
                        "dataTestId": "MobileNotification-all",
                        "key": "MobileNotification-all",
                        "name": "MobileNotification-all",
                        "suffix": undefined,
                        "title": Object {
                          "defaultMessage": "All new messages",
                          "id": "channel_notifications.MobileNotificationAllLabel",
                        },
                        "value": "all",
                      },
                      Object {
                        "dataTestId": "MobileNotification-mention",
                        "key": "MobileNotification-mention",
                        "name": "MobileNotification-mention",
                        "suffix": undefined,
                        "title": Object {
                          "defaultMessage": "Mentions, direct messages, and keywords only",
                          "id": "channel_notifications.MobileNotificationMentionLabel",
                        },
                        "value": "mention",
                      },
                      Object {
                        "dataTestId": "MobileNotification-none",
                        "key": "MobileNotification-none",
                        "name": "MobileNotification-none",
                        "suffix": undefined,
                        "title": Object {
                          "defaultMessage": "Nothing",
                          "id": "channel_notifications.MobileNotificationNothingLabel",
                        },
                        "value": "none",
                      },
                    ],
                  }
                }
                inputFieldValue="all"
                title={
                  Object {
                    "defaultMessage": "Notify me about…",
                    "id": "channel_notifications.NotifyMeTitle",
                  }
                }
              />
            </React.Fragment>
          </React.Fragment>
        }
        description={
          Object {
            "defaultMessage": "Notification alerts are pushed to your mobile device when there is activity in Mattermost.",
            "id": "channel_notifications.mobileNotificationsDesc",
          }
        }
        title={
          Object {
            "defaultMessage": "Mobile Notifications",
            "id": "channel_notifications.mobileNotificationsTitle",
          }
        }
        titleSuffix={
          <button
            className="channel-notifications-settings-modal__reset-btn"
            onClick={[Function]}
          >
            <RefreshIcon
              color="currentColor"
              size={14}
            />
            Reset to default
          </button>
        }
      />
      <div
        className="channel-notifications-settings-modal__divider"
      />
      <SectionCreator
        content={
          <React.Fragment>
            <CheckboxItemCreator
              handleChange={[Function]}
              inputFieldData={
                Object {
                  "dataTestId": "autoFollowThreads",
                  "name": "auto follow threads",
                  "title": Object {
                    "defaultMessage": "Automatically follow threads in this channel",
                    "id": "channel_notifications.checkbox.autoFollowThreadsTitle",
                  },
                }
              }
              inputFieldValue={false}
            />
          </React.Fragment>
        }
        description={
          Object {
            "defaultMessage": "When enabled, all new replies in this channel will be automatically followed and will appear in your Threads view.",
            "id": "channel_notifications.autoFollowThreadsDesc",
          }
        }
        title={
          Object {
            "defaultMessage": "Follow all threads in this channel",
            "id": "channel_notifications.autoFollowThreadsTitle",
          }
        }
      />
    </div>
  </main>
  <footer
    className="channel-notifications-settings-modal__footer"
  >
    <button
      className="channel-notifications-settings-modal__cancel-btn"
      onClick={[Function]}
    >
      <MemoizedFormattedMessage
        defaultMessage="Cancel"
        id="generic_btn.cancel"
      />
    </button>
    <button
      className="channel-notifications-settings-modal__save-btn"
      onClick={[Function]}
    >
      <MemoizedFormattedMessage
        defaultMessage="Save"
        id="generic_btn.save"
      />
    </button>
  </footer>
</Modal>
`;
